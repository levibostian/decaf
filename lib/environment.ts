import { exec } from "./exec.ts"
import * as log from "./log.ts"
import { AnyStepName } from "./steps/types/any-step.ts"
import envCi from "env-ci"

export interface Environment {
  getRepository(): { owner: string; repo: string }
  getBuild(): { buildUrl: string; buildId: string; currentBranch: string }
  getSimulatedMergeType(): "merge" | "rebase" | "squash"
  getEventThatTriggeredThisRun(): "push" | "pull_request" | "other"
  isRunningInPullRequest(): { baseBranch: string; targetBranch: string; prNumber: number } | undefined
  getCommandForStep({ stepName }: { stepName: AnyStepName }): string | undefined
  getGitConfigInput(): { name: string; email: string } | undefined
  setOutput({ key, value }: { key: string; value: string }): Promise<void>
  // A catch-all method to get inputs that dont match the other methods.
  getUserConfigurationOptions(): { failOnDeployVerification: boolean; makePullRequestComment: boolean }
}

export class EnvironmentImpl implements Environment {
  private readonly env: Record<string, string>
  private outputFileCache: Record<string, string>

  constructor() {
    // Example of the values in `this.env`:
    // {"isCi":true,"name":"GitHub Actions","service":"github","commit":"7d4aec10df2b2dcbe99643662beca90a24a8a81f","build":"15876053587","isPr":true,"branch":"alpha","prBranch":"refs/pull/68/merge","slug":"levibostian/new-deployment-tool","root":"/home/runner/work/new-deployment-tool/new-deployment-tool","pr":68}
    this.env = envCi()
    this.outputFileCache = {}
  }

  // Note: For pull requests, the return value is pretty useless. Example: "refs/pull/68/merge"
  getNameOfCurrentBranch(): string {
    if (this.env.isPr) return this.env.prBranch
    return this.env.branch
  }

  getBuild(): { buildUrl: string; buildId: string; currentBranch: string } {
    return {
      buildId: this.env.build,
      buildUrl: this.env.buildUrl,
      currentBranch: this.getNameOfCurrentBranch(),
    }
  }

  getRepository(): { owner: string; repo: string } {
    return {
      owner: this.env.slug.split("/")[0],
      repo: this.env.slug.split("/")[1],
    }
  }

  getSimulatedMergeType(): "merge" | "rebase" | "squash" {
    const githubActionInputKey = "simulated_merge_type"

    const simulateMergeType = this.getInput(githubActionInputKey)
    if (!simulateMergeType) {
      return "merge"
    }

    if (simulateMergeType !== "merge" && simulateMergeType !== "rebase" && simulateMergeType !== "squash") {
      log.error(
        `The value for the GitHub Actions input ${githubActionInputKey} is invalid. The value must be either "merge", "rebase", or "squash". The value provided was: ${simulateMergeType}`,
      )

      throw new Error()
    }

    return simulateMergeType
  }

  getEventThatTriggeredThisRun(): "push" | "pull_request" | "other" {
    if (this.env.isPr) return "pull_request"
    if (this.env.branch) return "push"

    return "other"
  }

  isRunningInPullRequest(): { baseBranch: string; targetBranch: string; prNumber: number } | undefined {
    if (!this.env.isPr) {
      return undefined
    }

    return {
      baseBranch: this.env.prBranch,
      targetBranch: this.env.branch,
      prNumber: Number(this.env.pr), // module provides this as a string, but we want it as a number.
    }
  }

  async setOutput({ key, value }: { key: string; value: string }): Promise<void> {
    await this.setOutputFile(key, value)

    if (this.env.service == "github") {
      await exec.run({
        command: `echo "${key}=${value}" >> "$GITHUB_OUTPUT"`,
        input: undefined,
      })
    }

    // Open to adding other CI services in the future, if they have a simple system for setting outputs.
  }

  getUserConfigurationOptions(): { failOnDeployVerification: boolean; makePullRequestComment: boolean } {
    return {
      failOnDeployVerification: this.getInput("fail_on_deploy_verification") === "true",
      makePullRequestComment: this.getInput("make_pull_request_comment") === "true",
    }
  }

  getGitConfigInput(): { name: string; email: string } | undefined {
    const gitConfigInput = this.getInput("git_config")
    if (!gitConfigInput || gitConfigInput.trim() === "") {
      return undefined
    }

    // Expect format: "Name <email>"
    const match = gitConfigInput.match(/^(.*)\s+<([^>]+)>$/)
    if (!match) {
      log.error(
        `The git_config input must be in the format "name <email>". The value provided was: ${gitConfigInput}`,
      )
      throw new Error()
    }

    const name = match[1].trim()
    const email = match[2].trim()

    return { name, email }
  }

  getCommandForStep({ stepName }: { stepName: string }): string | undefined {
    const command = this.getInput(stepName)
    if (!command) return undefined
    return command
  }

  private getInput(key: string): string {
    const val: string = Deno.env.get(`INPUT_${key.replace(/ /g, "_").toUpperCase()}`) || ""
    if (!val) {
      throw new Error(`The GitHub Actions input "${key}" is not set.`)
    }

    return val.trim()
  }

  // Writes output to a file as a way to support other CI services that are not github actions.
  private async setOutputFile(key: string, value: string): Promise<void> {
    const nameOfOutputFile = Deno.env.get("INPUT_OUTPUT_FILE")
    if (!nameOfOutputFile || nameOfOutputFile.trim() === "") return

    this.outputFileCache[key] = value

    await Deno.writeFile(
      nameOfOutputFile,
      new TextEncoder().encode(JSON.stringify(this.outputFileCache, null, 2)),
    )
  }
}
