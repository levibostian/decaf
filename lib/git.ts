import { exec } from "./exec.ts"
import { Exec } from "./exec.ts"
import { GitHubCommit } from "./github-api.ts"
import * as log from "./log.ts"

export interface Git {
  checkoutBranch: (
    { exec, branch, createBranchIfNotExist }: { exec: Exec; branch: string; createBranchIfNotExist: boolean },
  ) => Promise<void>
  merge: (
    { exec, branchToMergeIn, commitTitle, commitMessage, fastForward }: {
      exec: Exec
      branchToMergeIn: string
      commitTitle: string
      commitMessage: string
      fastForward?: "--no-ff" | "--ff-only"
    },
  ) => Promise<void>
  pull: ({ exec }: { exec: Exec }) => Promise<void>
  setUser: (
    { exec, name, email }: { exec: Exec; name: string; email: string },
  ) => Promise<void>
  squash: (
    { exec, branchToSquash, branchMergingInto, commitTitle, commitMessage }: {
      exec: Exec
      branchToSquash: string
      branchMergingInto: string
      commitTitle: string
      commitMessage: string
    },
  ) => Promise<void>
  rebase: (
    { exec, branchToRebaseOnto }: { exec: Exec; branchToRebaseOnto: string },
  ) => Promise<void>
  getLatestCommitsSince({ exec, commit }: { exec: Exec; commit: GitHubCommit }): Promise<GitHubCommit[]>
  getLatestCommitOnBranch({ exec, branch }: { exec: Exec; branch: string }): Promise<GitHubCommit>
  createLocalBranchFromRemote: ({ exec, branch }: { exec: Exec; branch: string }) => Promise<void>
  getCommits: ({ exec, branch }: { exec: Exec; branch: string }) => Promise<GitHubCommit[]>
}

const checkoutBranch = async (
  { exec, branch, createBranchIfNotExist }: { exec: Exec; branch: string; createBranchIfNotExist: boolean },
): Promise<void> => {
  await exec.run({
    command: `git checkout ${createBranchIfNotExist ? "-b " : ""}${branch}`,
    input: undefined,
  })
}

const merge = async (
  { exec, branchToMergeIn, commitTitle, commitMessage, fastForward }: {
    exec: Exec
    branchToMergeIn: string
    commitTitle: string
    commitMessage: string
    fastForward?: "--no-ff" | "--ff-only"
  },
): Promise<void> => {
  await exec.run({
    command: `git merge ${branchToMergeIn} -m "${commitTitle}" -m "${commitMessage}" ${fastForward || ""}`,
    input: undefined,
  })
}

const pull = async ({ exec }: { exec: Exec }): Promise<void> => {
  await exec.run({
    command: `git pull`,
    input: undefined,
  })
}

const setUser = async (
  { exec, name, email }: { exec: Exec; name: string; email: string },
): Promise<void> => {
  await exec.run({
    command: `git config user.name "${name}"`,
    input: undefined,
  })

  await exec.run({
    command: `git config user.email "${email}"`,
    input: undefined,
  })
}

// Squash all commits of a branch into 1 commit
const squash = async (
  { exec, branchToSquash, branchMergingInto, commitTitle, commitMessage }: {
    exec: Exec
    branchToSquash: string
    branchMergingInto: string
    commitTitle: string
    commitMessage: string
  },
): Promise<void> => {
  // We need to find out how many commits 1 branch is ahead of the other to find out how many unique commits there are.
  const { stdout } = await exec.run({
    command: `git rev-list --count ${branchMergingInto}..${branchToSquash}`,
    input: undefined,
  })

  const numberOfCommitsAheadOfBranchMergingInto = parseInt(stdout.trim())

  if (numberOfCommitsAheadOfBranchMergingInto === 0) {
    log.message(`Branches ${branchToSquash} and ${branchMergingInto} are already up to date. No commits to squash.`)
    return
  }

  // Now that we know how many commits are ahead, we can squash all of those commits into 1 commit.
  await exec.run({
    command: `git reset --soft HEAD~${numberOfCommitsAheadOfBranchMergingInto}`,
    input: undefined,
  })
  await exec.run({
    command: `git commit -m "${commitTitle}" -m "${commitMessage}"`,
    input: undefined,
  })
}

const rebase = async (
  { exec, branchToRebaseOnto }: { exec: Exec; branchToRebaseOnto: string },
): Promise<void> => {
  await exec.run({
    command: `git rebase ${branchToRebaseOnto}`,
    input: undefined,
  })
}

const getLatestCommitsSince = async (
  { exec, commit }: { exec: Exec; commit: GitHubCommit },
): Promise<GitHubCommit[]> => {
  const { stdout } = await exec.run({
    command: `git log --pretty=format:"%H|%s|%ci" ${commit.sha}..HEAD`,
    input: undefined,
  })

  return stdout.trim().split("\n").map((commitString) => {
    const [sha, message, dateString] = commitString.split("|")

    return { sha, message, date: new Date(dateString) }
  })
}

const getLatestCommitOnBranch = async (
  { exec, branch }: { exec: Exec; branch: string },
): Promise<GitHubCommit> => {
  const { stdout } = await exec.run({
    command: `git log -1 --pretty=format:"%H|%s|%ci" ${branch}`,
    input: undefined,
  })

  const [sha, message, dateString] = stdout.trim().split("|")

  return { sha, message, date: new Date(dateString) }
}

/**
 * Makes sure that we have a local branch that has all of the commits that the remote branch has.
 *
 * There are a lot of commands here, just to get a local branch of a remote branch. After many attempts to simplify this, we would hit many different errors.
 * I believe that complexity comes because we run this tool on a CI server where the git config might be different.
 * Running all of these commands and running each command by itself (example: not running `git checkout -b` to try and combine creating a branch and checking it out)
 * have given the most consistent results.
 */
const createLocalBranchFromRemote = async (
  { exec, branch }: { exec: Exec; branch: string },
): Promise<void> => {
  const currentBranchName = (await exec.run({
    command: `git branch --show-current`,
    input: undefined,
  })).stdout.trim()
  const doesBranchExist = (await exec.run({
    command: `git branch --list ${branch}`,
    input: undefined,
  })).stdout.trim() !== ""

  // Perform a fetch, otherwise you might get errors about origin branch not being found.
  await exec.run({
    command: `git fetch origin`,
    input: undefined,
  })

  // Only run if it doesn't exist locally. This is to avoid a error that crashes the tool: "fatal: a branch named '<branch-name>' already exists"
  if (!doesBranchExist) {
    await exec.run({
      command: `git branch --track ${branch} origin/${branch}`,
      input: undefined,
    })
  }

  // Checkout the branch so we can pull it.
  await exec.run({
    command: `git checkout ${branch}`,
    input: undefined,
  })

  // Pull the branch from the remote.
  // Adding --no-rebase to avoid an error that could happen when you run pull.
  // The error is: You have divergent branches and need to specify how to reconcile them.
  await exec.run({
    command: `git pull --no-rebase origin ${branch}`,
    input: undefined,
  })

  // Switch back to the branch we were on before.
  await exec.run({
    command: `git checkout ${currentBranchName}`,
    input: undefined,
  })
}

const getCommits = async (
  { exec, branch }: { exec: Exec; branch: string },
): Promise<GitHubCommit[]> => {
  const currentBranchName = (await exec.run({
    command: `git branch --show-current`,
    input: undefined,
  })).stdout.trim()

  await checkoutBranch({ exec, branch, createBranchIfNotExist: false })

  const { stdout } = await exec.run({
    command: `git log --pretty=format:"%H|%s|%ci"`,
    input: undefined,
  })

  const commits = stdout.trim().split("\n").map((commitString) => {
    const [sha, message, dateString] = commitString.split("|")

    return { sha, message, date: new Date(dateString) }
  })

  await checkoutBranch({ exec, branch: currentBranchName, createBranchIfNotExist: false })

  return commits
}

export const git: Git = {
  checkoutBranch,
  merge,
  pull,
  setUser,
  squash,
  rebase,
  getCommits,
  getLatestCommitsSince,
  getLatestCommitOnBranch,
  createLocalBranchFromRemote,
}
