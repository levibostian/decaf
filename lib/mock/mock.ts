import { stub, spy, GetParametersFromProp, GetReturnFromProp, Stub } from "jsr:@std/testing@1/mock";

/**
 * Example: 
 * ```
 * interface GitHubActions {
 *  isRunningInPullRequest(): Promise<boolean>
 * }
 * 
 * const githubActions: GitHubActions = mock()
 * when(githubActions, "isRunningInPullRequest", async () => true)
 * 
 * // optionally, get information about the mock: 
 * const isRunningInPullRequestMock = when(...)
 * assertEquals(isRunningInPullRequestMock.calls.length, 1)
 * ```
 */

export function when<T, Prop extends keyof T>(
  mockObject: T, 
  functionName: Prop, 
  thenCall: (this: T, ...args: GetParametersFromProp<T, Prop>) => GetReturnFromProp<T, Prop>): Stub<T, GetParametersFromProp<T, Prop>, GetReturnFromProp<T, Prop>> {
    return stub(mockObject, functionName, thenCall)
}

export function mock<T>(): T {
  const mockObject = {} as T;
  
  // for convenience, stub all functions with an empty implementation. So if a mock function doesn't need any logic, it can be used without needing to call `when` on it.
  for (const key in mockObject) {
    mockObject[key] = stub(mockObject, key) as unknown as T[Extract<keyof T, string>];
  }
  
  return mockObject;
}
