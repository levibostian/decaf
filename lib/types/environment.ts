import { GitHubCommit, GitHubRelease } from "../github-api.ts"

/*
  Each environment object contains information about the current state of the git repository and data that the tool has fetched/processed. Each step of the deployment may find this environment object useful to make decisions on what to do next.

  There are multiple different types of environment objects where each step of the process fetches/processes more data and adds it to the environment object.
*/

export interface GetLatestReleaseEnvironment {
  gitCurrentBranch: string
  gitRepoOwner: string
  gitRepoName: string
  testMode: boolean
}

export interface GetNextReleaseVersionEnvironment extends GetLatestReleaseEnvironment {
  lastRelease: GitHubRelease | null
  gitCommitsSinceLastRelease: GitHubCommit[]
}

export interface DeployEnvironment extends GetNextReleaseVersionEnvironment {
  nextVersionName: string
}
