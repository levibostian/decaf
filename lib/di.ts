import { defineStore } from "tiny-di"
import * as githubApi from "./github-api.ts"
import * as git from "./git.ts"
import { Environment, EnvironmentImpl } from "./environment.ts"

export const productionDiDefinition = defineStore()
  .add("git", git.impl)
  .add("github", githubApi.impl)
  .add("environment", (): Environment => new EnvironmentImpl())

export const productionDiGraph = productionDiDefinition

let diGraphToUseAtRuntime = productionDiGraph

export const overrideStore = (override: typeof productionDiGraph) => {
  // Override the production store with the provided store
  diGraphToUseAtRuntime = override
}

export const clearOverride = () => {
  // Clear any overrides and revert to the production store
  diGraphToUseAtRuntime = productionDiGraph
}

export const getGraph = () => diGraphToUseAtRuntime.finalize()
