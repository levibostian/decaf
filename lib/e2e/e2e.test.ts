/**
 * e2e tests, with a catch. To make these tests nice and fast, we do some mocking (but as little as possible).
 *
 * The idea is to run tests for scenarios that this tool is likely to encounter and make sure, from a product standpoint, the tool works as expected.
 */

import { assertEquals } from "@std/assert/equals"
import { EnvironmentStub, GitRemoteRepositoryMock, GitStub } from "./e2e-stubs.test.ts"
import { GitCommitFake } from "../types/git.test.ts"
import { GitCommit } from "../types/git.ts"
import * as e2eStepScript from "./e2e-step-script-helper.test.ts"
import { Environment } from "../environment.ts"
import { mock, when } from "../mock/mock.ts"
import { GitHubApi, GitHubPullRequest } from "../github-api.ts"
import { assertObjectMatch } from "@std/assert"
import { assertSnapshot } from "@std/testing/snapshot"
import * as di from "../di.ts"

Deno.test("when running a deployment, given CI only cloned 1 commit on current branch, expect to receive all parsed commits for branch", async (t) => {
  // when running on github actions with actions/checkout and it's default config, you will only have 1 checked out commit.
  const givenLocalCommits = new Map<string, GitCommit[]>([
    ["main", [new GitCommitFake({ message: "latest local commit on main branch" })]],
  ])
  const givenRemoteCommits = new Map<string, GitCommit[]>([
    ["main", [
      ...givenLocalCommits.get("main")!,
      new GitCommitFake({ message: "latest remote commit on main branch" }),
    ]],
  ])

  setupGitRepo({
    checkedOutBranch: "main",
    localCommits: givenLocalCommits,
    remoteCommits: givenRemoteCommits,
    remotePullRequests: [],
  })

  // this test only cares to assert the input data given for the getLatestRelease step
  // so, no need to setup more of the step scripts.
  e2eStepScript.setGetLatestReleaseStepOutput(null)

  await run({
    testMode: false,
  })

  // assert that the input data given to the step scripts is what we expect.
  // we expect that decaf got all of the commits from the remote repo.
  assertObjectMatch(e2eStepScript.getGetLatestReleaseInput(), {
    gitCurrentBranch: "main",
    testMode: false,
  })

  const actualMainBranchCommits = e2eStepScript.getGetLatestReleaseInput().gitCommitsAllLocalBranches["main"]
  assertEquals(actualMainBranchCommits.length, 2)
  await assertSnapshot(t, actualMainBranchCommits)
})

Deno.test("when running in test mode in a stacked pull request, expect the step scripts receive the simulated merge commits", async (t) => {
  // when running on github actions with actions/checkout and it's default config, you will only have 1 checked out commit.
  const mainBranchCommits = [new GitCommitFake({ message: "latest local commit on main branch" })]
  const feature1BranchCommits = [
    ...mainBranchCommits,
    new GitCommitFake({ message: "latest local commit on feature-1 branch" }),
  ]
  const feature2BranchCommits = [
    ...feature1BranchCommits,
    new GitCommitFake({ message: "latest local commit on feature-2 branch" }),
  ]

  const givenLocalCommits = new Map<string, GitCommit[]>([
    ["feature-2", feature2BranchCommits],
  ])

  const givenRemoteCommits = new Map<string, GitCommit[]>(
    [
      ["main", [
        new GitCommitFake({ message: "latest remote commit on main branch" }),
      ]],
      ["feature-1", feature1BranchCommits],
      ["feature-2", feature2BranchCommits],
    ],
  )

  setupGitRepo({
    checkedOutBranch: "feature-2",
    localCommits: givenLocalCommits,
    remoteCommits: givenRemoteCommits,
    remotePullRequests: [
      {
        prNumber: 124,
        targetBranchName: "feature-1",
        sourceBranchName: "feature-2",
        title: "PR for feature-2",
        description: "Description for feature-2",
      },
      {
        prNumber: 123,
        targetBranchName: "main",
        sourceBranchName: "feature-1",
        title: "PR for feature-1",
        description: "Description for feature-1",
      },
    ],
  })

  // this test only cares to assert the input data given for the getLatestRelease step
  // so, no need to setup more of the step scripts.
  e2eStepScript.setGetLatestReleaseStepOutput(null)

  await run({
    testMode: true,
  })

  // assert that the input data given to the step scripts is what we expect.
  // we expect that decaf got all of the commits from the remote repo.
  assertObjectMatch(e2eStepScript.getGetLatestReleaseInput(), {
    // expect that we are on the main branch when the step runs.
    gitCurrentBranch: "main",
    testMode: true,
  })

  const commitsReceivedInStep = e2eStepScript.getGetLatestReleaseInput().gitCommitsAllLocalBranches

  await assertSnapshot(t, commitsReceivedInStep)
})

// helper functions

let diGraph: typeof di.productionDiGraph

Deno.test.beforeEach(() => {
  di.clearOverride()

  diGraph = di.getGraph().createChild()
})

Deno.test.afterEach(() => {
  di.clearOverride()
})

const setupGitRepo = (
  { checkedOutBranch, localCommits, remoteCommits, remotePullRequests }: {
    checkedOutBranch: string
    localCommits: Map<string, GitCommit[]>
    remoteCommits: Map<string, GitCommit[]>
    remotePullRequests: GitHubPullRequest[]
  },
): { remoteRepository: GitRemoteRepositoryMock } => {
  currentBranchWhenTestStarts = checkedOutBranch

  const remoteRepository: GitRemoteRepositoryMock = {
    remoteBranches: remoteCommits,
    remoteTags: new Map(),
  }

  // Override the services with test implementations
  diGraph = diGraph.override("git", () => new GitStub({ currentBranch: checkedOutBranch, remoteRepo: remoteRepository, commits: localCommits }))

  githubApiMock = mock()
  when(githubApiMock, "getPullRequestStack", async (_args) => {
    return remotePullRequests
  })
  diGraph = diGraph.override("github", () => githubApiMock)

  return { remoteRepository }
}

let currentBranchWhenTestStarts: string
let environmentMock: Environment
let githubApiMock: GitHubApi = mock()

const run = async ({ testMode }: { testMode: boolean }) => {
  if (testMode) {
    environmentMock = new EnvironmentStub({
      commandToRunStubStepScript: e2eStepScript.getBashCommandToRunThisScript(),
      runFromPullRequest: {
        baseBranch: currentBranchWhenTestStarts,
        targetBranch: "feature",
        prNumber: 123,
        simulatedMergeType: "merge",
      },
    })
  } else {
    environmentMock = new EnvironmentStub({
      commandToRunStubStepScript: e2eStepScript.getBashCommandToRunThisScript(),
      runFromPush: {
        branch: currentBranchWhenTestStarts,
      },
    })
  }
  diGraph = diGraph.override("environment", () => environmentMock)

  // Finalize the DI graph with overrides
  di.overrideStore(diGraph)

  // Import the module and call its main function
  await import(`../../index.ts?t=${Date.now()}`)
}
