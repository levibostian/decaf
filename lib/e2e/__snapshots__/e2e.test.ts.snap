export const snapshot = {};

snapshot[`when running a deployment, given CI only cloned 1 commit on current branch, expect to receive all parsed commits for branch 1`] = `
[
  {
    abbreviatedSha: "abc123",
    author: {
      email: "test@example.com",
      name: "Test Author",
    },
    committer: {
      email: "test@example.com",
      name: "Test Committer",
    },
    date: "2021-01-01T00:00:00.000Z",
    filesChanged: [],
    isMergeCommit: false,
    isRevertCommit: false,
    message: "latest remote commit on main branch",
    messageLines: [
      "latest remote commit on main branch",
    ],
    parents: [],
    sha: "abc123",
    title: "latest remote commit on main branch",
  },
  {
    abbreviatedSha: "abc123",
    author: {
      email: "test@example.com",
      name: "Test Author",
    },
    committer: {
      email: "test@example.com",
      name: "Test Committer",
    },
    date: "2021-01-01T00:00:00.000Z",
    filesChanged: [],
    isMergeCommit: false,
    isRevertCommit: false,
    message: "latest local commit on main branch",
    messageLines: [
      "latest local commit on main branch",
    ],
    parents: [],
    sha: "abc123",
    title: "latest local commit on main branch",
  },
]
`;

snapshot[`when running in test mode in a stacked pull request, expect the step scripts receive the simulated merge commits 1`] = `
{
  "feature-1": [
    {
      abbreviatedSha: "abc123",
      author: {
        email: "test@example.com",
        name: "Test Author",
      },
      committer: {
        email: "test@example.com",
        name: "Test Committer",
      },
      date: "2021-01-01T00:00:00.000Z",
      filesChanged: [],
      isMergeCommit: false,
      isRevertCommit: false,
      message: "latest local commit on feature-1 branch",
      messageLines: [
        "latest local commit on feature-1 branch",
      ],
      parents: [],
      sha: "abc123",
      title: "latest local commit on feature-1 branch",
    },
    {
      abbreviatedSha: "abc123",
      author: {
        email: "test@example.com",
        name: "Test Author",
      },
      committer: {
        email: "test@example.com",
        name: "Test Committer",
      },
      date: "2021-01-01T00:00:00.000Z",
      filesChanged: [],
      isMergeCommit: false,
      isRevertCommit: false,
      message: "latest local commit on main branch",
      messageLines: [
        "latest local commit on main branch",
      ],
      parents: [],
      sha: "abc123",
      title: "latest local commit on main branch",
    },
  ],
  "feature-2": [
    {
      abbreviatedSha: "abc123",
      author: {
        email: "test@example.com",
        name: "Test Author",
      },
      committer: {
        email: "test@example.com",
        name: "Test Committer",
      },
      date: "2021-01-01T00:00:00.000Z",
      filesChanged: [],
      isMergeCommit: false,
      isRevertCommit: false,
      message: "latest local commit on feature-2 branch",
      messageLines: [
        "latest local commit on feature-2 branch",
      ],
      parents: [],
      sha: "abc123",
      title: "latest local commit on feature-2 branch",
    },
    {
      abbreviatedSha: "abc123",
      author: {
        email: "test@example.com",
        name: "Test Author",
      },
      committer: {
        email: "test@example.com",
        name: "Test Committer",
      },
      date: "2021-01-01T00:00:00.000Z",
      filesChanged: [],
      isMergeCommit: false,
      isRevertCommit: false,
      message: "latest local commit on feature-1 branch",
      messageLines: [
        "latest local commit on feature-1 branch",
      ],
      parents: [],
      sha: "abc123",
      title: "latest local commit on feature-1 branch",
    },
    {
      abbreviatedSha: "abc123",
      author: {
        email: "test@example.com",
        name: "Test Author",
      },
      committer: {
        email: "test@example.com",
        name: "Test Committer",
      },
      date: "2021-01-01T00:00:00.000Z",
      filesChanged: [],
      isMergeCommit: false,
      isRevertCommit: false,
      message: "latest local commit on main branch",
      messageLines: [
        "latest local commit on main branch",
      ],
      parents: [],
      sha: "abc123",
      title: "latest local commit on main branch",
    },
  ],
  main: [
    {
      abbreviatedSha: "c57ad98f",
      author: {
        email: "test@example.com",
        name: "Test User",
      },
      branch: "main",
      committer: {
        email: "test@example.com",
        name: "Test User",
      },
      date: "2025-09-21T11:50:37.569Z",
      filesChanged: [],
      isMergeCommit: true,
      isRevertCommit: false,
      message: "Merge pull request #123 from feature-1

",
      messageLines: [
        "Merge pull request #123 from feature-1",
        "",
        "",
      ],
      parents: [
        "abc123",
        "abc123",
      ],
      refs: [
        "main",
      ],
      sha: "e4dbe193cc95d800000000000000000000000000",
      title: "Merge pull request #123 from feature-1",
    },
    {
      abbreviatedSha: "abc123",
      author: {
        email: "test@example.com",
        name: "Test Author",
      },
      committer: {
        email: "test@example.com",
        name: "Test Committer",
      },
      date: "2021-01-01T00:00:00.000Z",
      filesChanged: [],
      isMergeCommit: false,
      isRevertCommit: false,
      message: "latest remote commit on main branch",
      messageLines: [
        "latest remote commit on main branch",
      ],
      parents: [],
      sha: "abc123",
      title: "latest remote commit on main branch",
    },
  ],
}
`;
